static-site/src/components/ListResources/IndividualResource.tsx

```diff
@@ -111,20 +129,20 @@ export const IndividualResource = ({data, isMobile}) => {
 
       <FlexRow>
 
-        <TooltipWrapper description={`Last known update on ${data.lastUpdated}`}>
-          <ResourceLinkWrapper onShiftClick={() => setModal(data)}>
-            <Name displayName={data.displayName} href={data.url} topOfColumn={topOfColumn}/>
+        <TooltipWrapper description={`Last known update on ${resource.lastUpdated}`}>
+          <ResourceLinkWrapper onShiftClick={() => setModalResource && setModalResource(resource as VersionedResource)}>
```

Comment:
I find amount of casting (`as` operator) excessive. It is not uncommon for projects in transition from js to ts to have some, in particularly uncertain places and where typed code needs to interact with yet untyped code. But upon a quick examination I found at least one serious typing issue which is being hidden by the cast.

Here we have a cast object `resource`, an instance of type `ResourceWithDisplayName` to  type `VersionedResource`. This is needed because `VersionedResource` the underlying type of the context, but `ResourceWithDisplayName` extends `Resource`, not `VersionedResource`.

The `Resource` interface contains all fields of the `VersionedResource` interface. So does it mean every `Resource` is also a `VersionedResource`? In this case this makes  little sense:

```ts
// types.ts
interface VersionedResource extends Resource
```

because it says "VersionedResource is a Resouce". So we just deduced that:

 - Resouce is VersionedResource
 - VersionedResource is Resouce

Meaning they are the same type.


I think (looking at names of the types) that the intent was different. Here is my proposal:

 - remove the 4 fields from the `Resource` and keep them only in `VersionedResource`: firstUpdated, dates, nVersions, updateCadence. Consider also moving versioned and lastUpdated (I don't know the meaning or concrete use, but from names they sound like they belong to something that's "versioned")

 - extend concrete resources from `VersionedResource`, rather than from `Resource`:

    ```ts
    export interface ResourceWithDisplayName extends VersionedResource {
    ```

 - obtain the following inheritance chain:

	```
	Resource -> VersionedResource -> ResourceWithDisplayName
	```

 - returning to the code snippet in review, the `resource`, which is still an instance of `ResourceWithDisplayName`, but now it is also an instance of `VersionedResource`, so no cast is needed.

I don't know if this proposal is compatible with the rest of the code, but my impression from the names of interfaces and fields is that this was the idea.

It might be worth to review all other occurrences of casting. It is better to keep casting to minimum, because it can lead to runtime errors if incompatible types are cast and then accessed incorrectly. It undermines the protection mechanism which is the whole point of adding ts. In C++ community they like to say "Casting is lying to your compiler, and compiler is your friend. You are lying to your friend who tries to help you." (citing in my words, but close to the original)

Reply:
@ivan-aksamentov thanks for flagging this - the `Resource`/`VersionedResource`/`ResourceWithDisplayName` thing was something I struggled with and could not come up with anything better on a Friday afternoon.

I went through all the type casts/assertions and replaced all but one: 7b24562d68ae048c949d0d1746b23ace37ea3941

Instead of extending `Resource` I figured it might be better to use it strictly and handle the optional attributes explicitly. What do you think of this approach?

Reply:
@victorlin Not sure. Time will tell. I don't quite like the proliferation of non-null assertions now. It's really the same as cast, but which is harder to find.

In order to decide the granullarity of typings, I'd usualy look at the origin of these objects (e.g. incoming HTTP request JSON, a file, a constant etc.). Where do they come from and in what shape?

If it comes as one monolithic object, then do one interface. If you find that in some places the null fields make no sense or become too tedious, then split the interface up and then combine the parts with `extend` and/or `&`. If you find that some function only wants a part of the object, then you could also split up.

One important hint:

If you's think about a program as of a black box with inputs and outputs:

```
input -> business logic -> output
```

then in TS you only need to type the input and the output - to make sure they have the shape that's expected on entry to and exit from the program. Within business logic, most types should be inferred by the compiler pretty well, so explicit types might not be needed at all. Notable exception is React, because you need props as all-in-one input, so you have to create new types there. For no reason really. But it's just a weird animal.

So if I'd convert a codebase to Typescript, I'd start with the inputs. For example, let's say for Auspice codebase, you could take Augur's JSON schema and generate TS types from it (I like using [quicktype](https://github.com/glideapps/quicktype)). And this will be like 80% of the typings for the whole program.

Reply:
I haven't written all that much TS, but I found that implementing something in TS involved a different mindset compared with a JS implementation. I found it very clarifying to be forced to define structures and not resort to conditional assertions & object modification which I tend to use a lot in pure JS. 

Looking at this code, there's no real difference in my JS implementation between `Resource`, `VersionedResource` and `ResourceWithDisplayName` objects - it's just that some have `undefined` values and we perform conditional checks on that in the consuming code.

This brings me to a more general point, and one I don't have a simple answer for. Converting JS to TS isn't a matter of simply adding type annotations to existing code, it is a process which should change the structure of the code as well. 

> If you's think about a program as of a black box with inputs and outputs... then in TS you only need to type the input and the output 

In this situation I think `<ListResources>` is a pretty good starting point - we have an input (JSON data sent from the server) and the output is react components.

Reply:
Thanks both for the input.

> There is a trade-off:
> 
> * (1) try to add typings with only minimal changes to the codebase (makeing it a refactoring)
> * (2) make typings sound (which requires fixes to the runtime behavior)

My goal with this PR is to do both sequentially - (1) with 8e5e5daab7b489ef5368d881f45f11930ef07de7 then (2) with 5151710b54c19ca4f86a3d7f039c5cadef2f0423.

---

> Looking at this code, there's no real difference in my JS implementation between `Resource`, `VersionedResource` and `ResourceWithDisplayName` objects - it's just that some have `undefined` values and we perform conditional checks on that in the consuming code.

This is now my sense of things too after removing the distinction in 7b24562d68ae048c949d0d1746b23ace37ea3941. I think it's cleaner this way and will squash most of that commit into 5151710b54c19ca4f86a3d7f039c5cadef2f0423.

---

The non-null assertions brought up in https://github.com/nextstrain/nextstrain.org/commit/7b24562d68ae048c949d0d1746b23ace37ea3941#r142683167 and https://github.com/nextstrain/nextstrain.org/commit/7b24562d68ae048c949d0d1746b23ace37ea3941#r142684106 are separate from the `Resource`/`VersionedResource`/`ResourceWithDisplayName` split - I'll think about and address those differently.